cpp/src/arrow/compute/kernels/hash.cc

```diff
@@ -415,5 +506,31 @@ Status DictionaryEncode(FunctionContext* ctx, const Datum& value, Datum* out) {
   return Status::OK();
 }
 
+const char kValuesFieldName[] = "Values";
+const char kCountsFieldName[] = "Counts";
+const int32_t kValuesFieldIndex = 0;
+const int32_t kCountsFieldIndex = 1;
+Status ValueCounts(FunctionContext* ctx, const Datum& value,
+                   std::shared_ptr<Array>* counts) {
+  std::unique_ptr<HashKernel> func;
+  RETURN_NOT_OK(GetValueCountsKernel(ctx, value.type(), &func));
+
+  // Calls return nothing for counts.
+  std::vector<Datum> unused_output;
+  std::shared_ptr<Array> uniques;
+  RETURN_NOT_OK(InvokeHash(ctx, func.get(), value, &unused_output, &uniques));
+
+  Datum value_counts;
+  RETURN_NOT_OK(func->FlushFinal(&value_counts));
+
+  auto data_type = std::make_shared<StructType>(std::vector<std::shared_ptr<Field>>{
+      std::make_shared<Field>("Values", uniques->type()),
+      std::make_shared<Field>("Counts", int64())});
+  *counts = std::make_shared<StructArray>(
+      data_type, uniques->length(),
+      std::vector<std::shared_ptr<Array>>{uniques, MakeArray(value_counts.array())});
```

Comment:
We should come up with a more elegant API for constructing StructArray https://issues.apache.org/jira/browse/ARROW-4788